src/openzaak/utils/tasks.py

```diff
@@ -0,0 +1,765 @@
+
+    try:
+        instance.save(update_fields=updated_fields)
+    except DatabaseError as e:
+        logger.critical(
+            f"Unable to save import state due to database error: {str(e)}"
+        )
+
+
+def _finish_batch(
+    import_instance: Import,
+    batch: list[DocumentRow],
+) -> None:
+    batch_number = import_instance.get_batch_number(len(batch))
+    _processed, _fail_count, _success_count = _get_batch_statistics(batch)
+
+    import_instance.processed = import_instance.processed + _processed
+    import_instance.processed_succesfully = (
+        import_instance.processed_succesfully + _success_count
+    )
+    import_instance.processed_invalid = (
+        import_instance.processed_invalid + _fail_count
+    )
+
+    try:
+        import_instance.save(
+            update_fields=["processed", "processed_succesfully", "processed_invalid"]
+        )
+    except DatabaseError as e:
+        logger.critical(
+            f"Unable to save batch statistics for batch {batch_number} due to database "
+            f"error: {str(e)}"
+        )
+
+    logger.info(f"Writing batch number {batch_number} to report file")
+    _write_to_file(import_instance, batch)
+
+
+
+LOCK_EXPIRE = 60 * (60 * 24)  # 24 hours
+
+
+@contextmanager
+def task_lock(lock_id, oid):
+    timeout_at = monotonic() + LOCK_EXPIRE - 3
+    logger.info(f"Lock id {lock_id}:{oid} cache added.")
+    status = cache.add(lock_id, oid, LOCK_EXPIRE)
+    try:
+        yield status
+    finally:
+        if monotonic() < timeout_at and status:
+            logger.warning(f"Lock id {lock_id}:{oid} cache deleted")
+            cache.delete(lock_id)
+
+
+# Note that this not will not work with per process caches (e.g LocMemCache)
+def task_locker(func):
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        instance = args[0]
+        lock_id = f"{instance.name}_lock"
+        with task_lock(lock_id, instance.app.oid) as acquired:
+            if acquired:
+                return func(*args, **kwargs)
+        logger.warning(f"Task {instance.name} already running, dispatch ignored...")
+
+    return wrapped
+
+
+# TODO: make this more generic?
+# TODO: expose `batch_size` through API?
+@celery_app.task(bind=True)
+@task_locker
+def import_documents(self, import_pk: int, batch_size=500) -> None:
+    import_instance = Import.objects.get(pk=import_pk)
+
+    file_path = import_instance.import_file.path
+
+    import_instance.total = _get_total_count(file_path)
+    import_instance.started_on = timezone.now()
+    import_instance.status = ImportStatusChoices.active
+    import_instance.save(update_fields=["total", "started_on", "status"])
+
+    batch: list[DocumentRow] = []
+
+    zaak_uuids = {str(uuid): id for uuid, id in Zaak.objects.values_list("uuid", "id")}
+
+    for row_index, row in _get_csv_generator(file_path):
+        if row_index == 1:  # skip the header row
+            continue
+
+        if len(batch) % batch_size == 0:
```

Comment:
Could we use celery chunks (https://docs.celeryq.dev/en/stable/userguide/canvas.html#chunks) instead of creating batches manually?